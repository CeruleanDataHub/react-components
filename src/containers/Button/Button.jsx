import PropTypes from "prop-types";
import React, { forwardRef } from "react";
import styled from "styled-components";
import { Icon } from "../Icon/Icon";

const StyledButton = styled.button`
  font-family: inherit;
  background: transparent;
  color: ${({ color }) => color};
  font-size: 1em;
  padding: 0.25em 1.5em;
  border: 1px solid ${({ color }) => color};
  border-radius: 4px;
  cursor: pointer;
`;

const StyledIcon = styled(Icon)`
  margin: 0 2px;
`

export const Button = forwardRef(
  ({ children, className, onClick, color, type, as, icon }, ref) => (
    <StyledButton data-button-test className={className} color={color} onClick={onClick} type={type} as={as} ref={ref}>
      { icon && <Icon name={icon} as={StyledIcon} data-button-icon-test /> }

      {children}
    </StyledButton>
  )
);

Button.propTypes = {
  /** Children of the button */
  children: PropTypes.node,
  className: PropTypes.string,
  /** Handler function */
  onClick: PropTypes.func.isRequired,
  /** Color of the button */
  color: PropTypes.string,
  /** Button type */
  type: PropTypes.string,
  /** Extend styles, property is from styled components */
  as: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.string,
    PropTypes.shape({ render: PropTypes.func.isRequired })
  ]),
  icon: PropTypes.string,
};

Button.defaultProps = {
  children: "",
  className: "",
  color: "",
  type: "button",
  as: null,
  icon: "",
};
