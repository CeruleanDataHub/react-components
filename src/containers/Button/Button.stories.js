import { action } from "@storybook/addon-actions";
import React from "react";

import { Button } from "./Button";

export default {
  title: "Button",
  component: Button
};

const colors = [
  "crimson",
  "tomato",
  "khaki",
  "mediumseagreen",
  "darkcyan",
  "steelblue",
  "magenta",
  "palevioletred"
];

export const ButtonStory = () => (
  <Button text="Button" color="blue" onClick={action("button clicked")} />
);

export const ButtonColorsStory = () =>
  colors.map(color => (
    <span style={{ margin: "0.5em" }} key={color}>
      <Button text="Button" color={color} onClick={action("button-clicked")} />
    </span>
  ));
