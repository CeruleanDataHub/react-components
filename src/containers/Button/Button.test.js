import { shallow } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";
import styled from "styled-components";

import { Button } from "./Button";

const colors = ["crimson", "tomato", "khaki"];

describe("Button", () => {
  it("should render Button", () => {
    const component = renderer.create(<Button />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render Button with props", () => {
    const component = renderer.create(<Button text="Button" color="blue" />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should accept 'as' property", () => {
    const CustomButton = styled.button`
      background: black;
    `;

    const component = renderer.create(
      <Button text="Button" color="blue" as={CustomButton} />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("Should render Button color based on props", () => {
    colors.map(color =>
      it("renders with correct color", () => {
        const component = shallow(<Button text="Button" color={color} />);
        expect(component).toHaveStyleRule("color", color);
        expect(component).toHaveStyleRule("border", `1px solid ${color}`);
      })
    );
  });
});
