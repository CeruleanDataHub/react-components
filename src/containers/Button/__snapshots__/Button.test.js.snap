// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should accept 'as' property 1`] = `
.c1 {
  background: transparent;
  color: blue;
  font-size: 1em;
  padding: 0.25em 1.5em;
  border: 1px solid blue;
  border-radius: 4px;
  cursor: pointer;
}

.c0 {
  background: black;
}

<button
  className="c0 c1"
  color="blue"
  onClick={[Function]}
>
  Button
</button>
`;

exports[`Button should render Button 1`] = `
.c0 {
  background: transparent;
  font-size: 1em;
  padding: 0.25em 1.5em;
  border: 1px solid;
  border-radius: 4px;
  cursor: pointer;
}

<button
  className="c0"
  color=""
  onClick={[Function]}
>
  
</button>
`;

exports[`Button should render Button with props 1`] = `
.c0 {
  background: transparent;
  color: blue;
  font-size: 1em;
  padding: 0.25em 1.5em;
  border: 1px solid blue;
  border-radius: 4px;
  cursor: pointer;
}

<button
  className="c0"
  color="blue"
  onClick={[Function]}
>
  Button
</button>
`;
